metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema_gen.py

```diff
@@ -0,0 +1,1096 @@
+
+        db_views: Dict[str, List[BigqueryView]] = {}
+        db_snapshots: Dict[str, List[BigqueryTableSnapshot]] = {}
+        project_id = bigquery_project.id
+
+        q: "queue.Queue[MetadataWorkUnit]" = queue.Queue(maxsize=100)
+
+        def _process_schema_worker(bigquery_dataset: BigqueryDataset) -> None:
+            if not is_schema_allowed(
+                self.config.dataset_pattern,
+                bigquery_dataset.name,
+                project_id,
+                self.config.match_fully_qualified_names,
+            ):
+                self.report.report_dropped(f"{bigquery_dataset.name}.*")
+                return
+            try:
+                # db_tables, db_views, and db_snapshots are populated in the this method
+                for wu in self._process_schema(
+                    project_id, bigquery_dataset, db_tables, db_views, db_snapshots
+                ):
+                    q.put(wu)
+            except Exception as e:
+                error_message = f"Unable to get tables for dataset {bigquery_dataset.name} in project {project_id}, skipping. Does your service account has bigquery.tables.list, bigquery.routines.get, bigquery.routines.list permission? The error was: {e}"
+                if self.config.is_profiling_enabled():
+                    error_message = f"Unable to get tables for dataset {bigquery_dataset.name} in project {project_id}, skipping. Does your service account has bigquery.tables.list, bigquery.routines.get, bigquery.routines.list permission, bigquery.tables.getData permission? The error was: {e}"
+                trace = traceback.format_exc()
+                logger.error(trace)
+                logger.error(error_message, exc_info=e)
+                self.report.report_failure(
+                    "metadata-extraction",
+                    f"{project_id}.{bigquery_dataset.name} - {error_message} - {trace}",
+                )
+
+        with concurrent.futures.ThreadPoolExecutor(
+            max_workers=SCHEMA_PARALLELISM
+        ) as executor:
+            futures = []
+            for bq_dataset in bigquery_project.datasets:
+                f = executor.submit(_process_schema_worker, bq_dataset)
+                futures.append(f)
+
+            # Read from the queue and yield the work units until all futures are done.
+            while True:
+                if not q.empty():
+                    while not q.empty():
+                        yield q.get_nowait()
+                else:
+                    try:
+                        yield q.get(timeout=0.2)
+                    except queue.Empty:
+                        pass
+
+                # Filter out the done futures.
+                futures = [f for f in futures if not f.done()]
+                if not futures:
+                    break
+
+        # Yield the remaining work units. This theoretically should not happen, but adding it just in case.
+        while not q.empty():
+            yield q.get_nowait()
```

Comment:
does it make sense to pull some of this out into a helper method or something?

Reply:
ideally reused across bq and snowflake

Reply:
Yes

Reply:
Done